Assetgraph propagate symbols

mischnic opened the pull request:
# ↪️ Pull Request

Supersedes https://github.com/parcel-bundler/parcel/pull/4841
Closes #4565

Closes T-675

- After building the asset graph, propagate symbols in two passes (up & down). This is faster than doing it while building the asset graph because the bookkeeping (many nodes are visited multiple times) brings too much overhead.

- Use `asset.usedSymbols` information to:
	- exclude JS assets in the JSPackager
	- When exposing an asset to a different bundle (so shared bundles are shaken as well)
- Use `dependency.usedSymbols` information to:
	- Skip unused subgraphs in the bundler. So bundles for unused imports aren't even created and CSS imports from excluded assets are excluded as well (but `sideEffects: ["*.css"]` is usually required to get any CSS at all).
		- **This condition (currently hardcoded in the BundleGraph) should really be exposed to the bundler!**
	- This essentially behaves just like deferring the assets, but still excludes these files if deferring deopts (chained reexports / namespace)
- We now need to check for missing exports in the AssetGraphBuilder, because when importing a missing symbol would mean that no symbols in that subgraph are actually used and the subgraph is removed entirely. The packager wouldn't be able to differentiate between "asset is missing because an import was wrong" and "asset is missing because it was unused & excluded". **So resolving symbols essentially happens not when replacing the used import but when resolving the dependency, this might cause problems for TS's "phantom" imports?** This could be alleviated by not adding a symbol to the dependency when hoisting if it's not actually referenced in the asset. 

Follow up tasks:

1. Tree shaking dynamic imports
2. Somehow operate on symbols cross assets (e.g. skip TS type reexports)
3. actually do the symbol resolution in the graph (not sure how slower this would be). Could catch conflicting reexports (to be spec compliant)



## ✔️ PR Todo

- [x] Update two tests in `html.js`
- [x] Test TS imports? (Stray type imports that weren't removed during transpilation)
- [x] make `this.propagateSymbols()` incremental (only update when assets changed)

## Complete Benchmark 

```
v2, cold cache
	191.35s user 28.42s system 374% cpu 58.626 total
    196.22s user 30.34s system 372% cpu 1:00.84 total
 	178.72s user 27.41s system 374% cpu 55.096 total
 gdu -sB1 dist
	16654336	dist


v2, no changes:
	9.48s user 1.45s system 213% cpu 5.123 total
	11.49s user 1.80s system 213% cpu 6.234 total
	10.27s user 1.59s system 209% cpu 5.653 total

v2, cold cache, minify:
	271.32s user 34.83s system 328% cpu 1:33.24 total
 	270.44s user 33.86s system 331% cpu 1:31.82 total
  gdu -sB1 dist
	5873664	dist

-------

propagate, cold cache:
	190.78s user 28.86s system 380% cpu 57.648 total
	182.22s user 28.42s system 370% cpu 56.867 total
	180.72s user 28.06s system 375% cpu 55.613 total
 gdu -sB1 dist
	16289792	dist

propagate, no changes:
	9.76s user 1.49s system 211% cpu 5.310 total
	10.63s user 1.62s system 208% cpu 5.880 total
	10.38s user 1.59s system 208% cpu 5.752 total

propagate, cold cache, minify:
	254.50s user 31.51s system 304% cpu 1:33.88 total
	251.15s user 30.40s system 343% cpu 1:21.97 total
 	248.30s user 31.13s system 338% cpu 1:22.67 total
	243.36s user 30.87s system 344% cpu 1:19.71 total
 gdu -sB1 dist
	5672960	dist
```

## Benchmark (while determining which strategy to use)

Times in ms

`ak-editor$ parcel2 build . --no-minify --no-source-maps`

baseline: **not incremental:**

```
main   : 700, 834, 683, 757
runtime:  30,  26,  22,  25
```

baseline: **without the topological ordering:**

```
main   : 309, 361
runtime:  18,  21
```


current: **incremental, cold cache:**

main
```
down: 249, 246, 213, 227
up  :  92,  64,  65,  69
```

runtime
```
down:   6,  17,   7,   4
up  :   2,   2,   3,   2
```

current: **incremental, no change**

main
```
down: 40, 38, 39
up  : 61, 58, 58
```

"up" is slow because of getIncomingDependencies

mischnic commented on packages/bundlers/default/src/DefaultBundler.js:
```diff
@@ -66,6 +66,11 @@ export default new Bundler({
         let assets = bundleGraph.getDependencyAssets(dependency);
         let resolution = bundleGraph.getDependencyResolution(dependency);
 
+        if (bundleGraph.isDependencySkipped(node.value)) {
```
Since this is done implicitly in `bundleGraph.addAssetGraph`, should this check be done in `bundleGraph.traverse` itself and the callback never be called?

mischnic replied:
Actually, I've had to add this kind of check to `addAssetGraphToBundle` (so inside `BundleGraph`). But ideally this would be configureable by the caller (regarding https://github.com/parcel-bundler/parcel/pull/4983)

mischnic replied:
> Are there any cases where we would want to traverse over skipped nodes?

Same as https://github.com/parcel-bundler/parcel/pull/4983, it's if the plugin author's definition of "skip" and Parcel's opinion diverge (= there is some other decision factor besides symbols).

I've now added an optional callback to `addAssetGraphToBundle` that can be used to control this. I think we should just keep the explicit check in the bundler traversal.

mischnic commented on packages/core/types/index.js:
```diff
@@ -271,33 +271,55 @@ export type SourceLocation = {|
+  exportSymbols(): Iterable<Symbol>;
 }
-export interface MutableSymbols extends Symbols {
-  // Static analysis bailed out
```
In hindsight, having cleared symbols on an asset doesn't make any sense.
A symbol `*:$id$exports` can be used as a fallback instead (this exists on CommonJS and not on ES6 modules)

devongovett commented on packages/shared/scope-hoisting/src/utils.js:
```diff
@@ -314,11 +311,143 @@ export function convertBabelLoc(loc: ?BabelSourceLocation): ?SourceLocation {
   };
 }
+
+export function getExportNamespaceExpression(
```
Are we not keeping the namespaces generated by hoist anymore?

mischnic replied:
If only I could remember why I did this 🙈

mischnic replied:
So I think the reason here was that exports in shared bundles wouldn't be shaken because the namespace objects always contains all exports, not just the used ones. 

So if you have a shared bundle that contains this asset
```js
export const a = 2;
export const b = 4;
```

and `b` is never actually used, this synthetic namespace objects contains `{a: $id$export$a}` so that `b` can be `shake()`n.

mischnic commented on packages/shared/scope-hoisting/src/concat.js:
```diff
@@ -195,87 +192,10 @@ function parse(code, sourceFilename) {
-      // Don't exclude assets that was imported as a wildcard
-      d.symbols.hasExportSymbol('*'),
-    )
+    bundleGraph.getUsedSymbolsAsset(asset).size == 0
```
There are two ways an asset(subgraph) can be excluded:
- the whole subgraph can be skipped if all dependencies to it are unused (= no used symbols on `dependency.usedSymbols`), so this would be some unused import/reexport
- a single asset can be skipped if none of its symbols are used (`asset.usedSymbols`), so this would be a file which has only reexports and can be excluded while its children are still needed

mischnic commented on packages/core/integration-tests/test/scope-hoisting.js:
```diff
@@ -2484,7 +3317,7 @@ describe('scope hoisting', function() {
   });
 
-  it('deduplicates shared sibling assets between bundle groups', async () => {
+  it.skip('deduplicates shared sibling assets between bundle groups', async () => {
```
🤷‍♂️ it passes

mischnic commented on packages/core/integration-tests/test/scope-hoisting.js:
```diff
@@ -580,553 +607,1356 @@ describe('scope hoisting', function() {
+        new Set(['c']),
+      );
+
+      // TODO concat order is wrong?
```
```
  const $ed3b52c582d5cb17749eed21a602a595$export$c = "c:" + $d668fa94ea609b744162b71372ed5e61$export$b;
                                                            ^

ReferenceError: Cannot access '$d668fa94ea609b744162b71372ed5e61$export$b' before initialization
      at evalmachine.<anonymous>:4:61
```

mischnic commented on packages/shared/scope-hoisting/src/formats/commonjs.js:
```diff
@@ -441,39 +438,51 @@ export function generateExports(
+                );
+              }
             }
+          } else {
```
I've added a comment:
```js
            // `getExportedSymbols` can return `$id$import$foo` symbols so that cross-bundle imports
            // are resolved correctly. There is no binding in that case.
```

mischnic commented on packages/shared/scope-hoisting/src/formats/esmodule.js:
```diff
@@ -325,9 +324,85 @@ export function generateExports(
+      t.variableDeclaration('var', declarations),
+      t.exportNamedDeclaration(null, exportedIdentifiersBailoutSpecifiers),
+    ]);
+    programPath.scope.crawl();
```
Apparently I fixed this already and only removed that call for the commonjs format

mischnic commented on packages/shared/scope-hoisting/src/formats/esmodule.js:
```diff
@@ -325,9 +324,85 @@ export function generateExports(
+    }
+  }
+
+  if (exportedIdentifiersBailout.size > 0) {
```
A esmodule bundle could reexport something from an asset in different bundle, but I don't think this can actually happen at the moment (and there are apparently no tests for this). Should I just remove that branch with an assertion for now?

devongovett replied:
Either that or add a comment?

mischnic replied:
I've commented out the code and added back the error/assertion that was thrown previously in this situation.

mischnic commented on packages/core/core/src/AssetGraph.js:
```diff
@@ -134,6 +144,15 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
   removeNode(node: AssetGraphNode): void {
     this.hash = null;
     this.onNodeRemoved && this.onNodeRemoved(node);
+    // FIXME cleanup, this needs to mark all connected nodes that doesn't become orphaned
```
I don't know how to solve this differently, so I've removed the todo comment and added some reasoning.

mischnic commented:
Still to do:

```js
import a from "url:./b.txt";
import b from "bundle-text:./b.js";
import c from "data-url:./b.txt";
```

all fail because the pipeline isn't taken into account (and these are handled by packagers/runtimes).

Furthermore, I think `asset.symbols.isCleared` should probably be brought back to be able to opt out of this system...

mischnic commented:
I've simplified the JS codegen part: instead of generating a namespace object expression when exposing an asset in a bundle (which lead to horrible code and also made these exports non-live)
```js
function $id$init() {
	return $parcel$defineInterop({
		foo: $id$export$foo,
		bar: $otherid$export$thing
	});
}
```

I've changed the hoist transformer to also emit `$parcel$reexport` (which I renamed to `$parcel$export`) for normal ESM export (not reexport), meaning that:
```js
var $id$exports = {};
var $id$export$foo = "xyz";
$parcel$export($id$exports, 'foo', function(){ return $id$export$foo; }); //previously: $id$exports.foo = $id$export$foo

function $id$init() {
	return $id$exports;
}
```

These calls can now be easily removed by checking for every `$parcel$export($id$exports, 'foo'` whether the symbol `'foo'` exported by asset `id` is used.

mischnic commented on packages/core/core/src/AssetGraph.js:
```diff
@@ -359,6 +401,8 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
   }
 
   resolveAsset(assetNode: AssetNode, dependentAssets: Array<Asset>) {
+    // $FlowFixMe FIXME cleanup, this is because of the magic of addNode (called by replaceNodesConnectedTo)
+    assetNode = this.getNode(assetNode.id);
     let depNodes = [];
     let depNodesWithAssets = [];
```
...
